docs/tools/testing/unit_testing.md

```diff
@@ -1,27 +1,275 @@
 
-#### Building, running and writing tests
+### Writing unit tests
+
+Each unit test suite requires at least two files to be created:
+1. C++ unit test source file (e.g. `test_NetworkInterface.cpp`)
+2. CMake module file for unit test definition (`unittest.cmake`)
+
+A unit test definition file `unittest.cmake` requires variables to be set for a test to be configured. File source paths in `unittest.cmake` files need to be relative to the unit test folder and `CMakeLists.txt`.
+
+* **TEST_SUITE_NAME** - Unique identifier for the test suite. Use naming convention *PATH-TO-THE-TESTABLE-FILE* e.g. *features-netsocket-InternetSocket*
+* **unittest-includes** - Include paths for headers needed to build the tests in addition to the base include paths listed in CMakeLists.txt. Optional.
+* **unittest-sources** - Mbed OS source files and stubs included for the build.
+* **unittest-test-sources** - Unit test source files.
+
+##### Creating unit tests files with Mbed CLI
+
+To learn how to generate test files, please see [a handbook section about using unit tests with Mbed CLI](https://os.mbed.com/docs/latest/tools/test-and-debug.html#unit-testing). The generator can only create the files required for a unit test. It does not write unit tests automatically nor does it handle source dependencies.
+
+##### Create files manually
+
+For example to create a unit test for `rtos/Semaphore.cpp`:
+
+1. Create a directory for unit test files in `UNITTESTS/rtos/Semaphore`.
+2. Create a test definition file `UNITTESTS/rtos/Semaphore/unittest.cmake` with the following content:
+```
+set(TEST_SUITE_NAME "rtos-Semaphore")
+
+set(unittest-sources
+	stubs/mbed_assert.c
+	../rtos/Semaphore.cpp
+)
+
+set(unittest-test-sources
+	rtos/Semaphore/test_Semaphore.cpp
+)
+```
+3. Create a test source file `UNITTESTS/rtos/Semaphore/test_Semaphore.cpp` with the following content:
+```
+#include  "gtest/gtest.h"
+#include  "rtos/Semaphore.h"
+
+static osStatus_t retval = osOK;
+static uint32_t count = 0;
+
+// Test stubs
+osStatus_t osSemaphoreAcquire(osSemaphoreId_t semaphore_id, uint32_t timeout)
+{
+    return retval;
+}
+osStatus_t osSemaphoreDelete(osSemaphoreId_t semaphore_id)
+{
+    return retval;
+}
+osStatus_t osSemaphoreRelease(osSemaphoreId_t semaphore_id)
+{
+    return retval;
+}
+uint32_t osSemaphoreGetCount(osSemaphoreId_t semaphore_id)
+{
+    return count;
+}
+osSemaphoreId_t osSemaphoreNew(uint32_t max_count, uint32_t initial_count, const osSemaphoreAttr_t *attr)
+{
+    return (void *)&count; // Just a dymmy reference
+}
+
+class  TestSemaphore : public  testing::Test {
+protected:
+    rtos::Semaphore *sem;
+
+    virtual  void  SetUp()
+    {
+        sem = new rtos::Semaphore();
+    }
+
+    virtual  void  TearDown()
+    {
+        delete sem;
+    }
+};
+
+TEST_F(TestSemaphore, constructor)
+{
+    EXPECT_TRUE(sem);
+}
+```
+
+### Building and running unit tests
+
+#### Testing with Mbed CLI
+
```

Comment:
Relating to my previous comment: Especially since all we give about Mbed CLI is some background and a link, I see no point in making people look for the CLI content on this page again.